src/execution/operator/csv_scanner/encode/csv_decoder.cpp

```diff
@@ -0,0 +1,97 @@
+#include "duckdb/execution/operator/csv_scanner/encode/csv_decoder.hpp"
+#include "duckdb/common/exception.hpp"
+#include "duckdb/main/config.hpp"
+
+namespace duckdb {
+
+void CSVEncoderBuffer::Initialize(idx_t encoded_size) {
+	encoded_buffer_size = encoded_size;
+	encoded_buffer = std::unique_ptr<char[]>(new char[encoded_size]);
+}
+
+char *CSVEncoderBuffer::Ptr() const {
+	return encoded_buffer.get();
+}
+
+idx_t CSVEncoderBuffer::GetCapacity() const {
+	return encoded_buffer_size;
+}
+
+idx_t CSVEncoderBuffer::GetSize() const {
+	return actual_encoded_buffer_size;
+}
+
+void CSVEncoderBuffer::SetSize(const idx_t buffer_size) {
+	D_ASSERT(buffer_size <= encoded_buffer_size);
+	actual_encoded_buffer_size = buffer_size;
+}
+
+bool CSVEncoderBuffer::HasDataToRead() const {
+	return cur_pos < actual_encoded_buffer_size;
+}
+
+void CSVEncoderBuffer::Reset() {
+	cur_pos = 0;
+	actual_encoded_buffer_size = 0;
+}
+
+CSVDecoder::CSVDecoder(DBConfig &config, const string &enconding_name_p, idx_t buffer_size) {
+	encoding_name = StringUtil::Lower(enconding_name_p);
+	auto function = config.GetDecodeFunction(encoding_name);
+	if (!function) {
+		auto loaded_encodings = config.GetLoadedDecodeFunctionNames();
+		std::ostringstream error;
+		error << "The CSV Reader does not support the encoding: \"" << enconding_name_p << "\"\n";
+		error << "The currently supported encodings are: " << '\n';
+		for (auto &encoding_name : loaded_encodings) {
+			error << "*  " << encoding_name << '\n';
+		}
+		throw InvalidInputException(error.str());
+	}
+	// Let's enforce that the encoded buffer size is divisible by 2, makes life easier.
+	idx_t encoded_buffer_size = buffer_size / function->GetRatio();
```

Comment:
I'm not entirely following the code here - should we not be multiplying the `buffer_size` by the ratio? For example, for `latin1`, in the worst case we get a 2X blow-up (when all bytes are `>128`, e.g. `ÀÀÀÀÀÀÀÀÀ`). The division here seems a bit odd?

Reply:
This is basically what this is doing but the other way around.
It was necessary before I was actually caching the buffer, I can remove that logic now.

The only thing is that the encoded buffer size must be an even number, otherwise the conversion of UTF-16 becomes more finicky

Reply:
Ah yeah, that makes sense. Maybe we want to have some encoding state or so? That way the encoding itself can keep track of such extra information